File: solutions/202-happy-number/solution.py
class Solution:
    def is_happy(self, n: int) -> bool:
        seen: set[int] = set()

        while n != 1 and n not in seen:
            seen.add(n)
            n = sum(int(d) * int(d) for d in str(n))

        return n == 1


if __name__ == "__main__":
    tests = [
        (19, True),
        (2, False),
        (1, True),
        (7, True),  # 7 → 49 → 97 → 130 → 10 → 1
    ]
    sol = Solution()
    for n, expected in tests:
        result = sol.is_happy(n)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | n={n} => {result} (expected {expected})")

File: solutions/202-happy-number/solution.go
package main

import "fmt"

func squareSum(n int) int {
	sum := 0
	for n > 0 {
		d := n % 10
		sum += d * d
		n /= 10
	}
	return sum
}

func isHappy(n int) bool {
	seen := make(map[int]bool)

	for n != 1 && !seen[n] {
		seen[n] = true
		n = squareSum(n)
	}
	return n == 1
}

func main() {
	tests := []struct {
		n        int
		expected bool
	}{
		{19, true},
		{2, false},
		{1, true},
		{7, true},
	}

	for _, tc := range tests {
		result := isHappy(tc.n)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | n=%d => %v (expected %v)\n", status, tc.n, result, tc.expected)
	}
}
